File: 0022-generate-parentheses/solution.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<String>();
        int openUsed = 0;
        int closedUsed = 0;
        String curr = "";

        generateCombs(openUsed, closedUsed, n, ans, curr);
        return ans;
    }

    public void generateCombs(int openUsed, int closedUsed, int n, List<String> ans, String curr) {
        if(curr.length() == 2*n) {
            ans.add(curr);
            return;
        }
        if(openUsed < n) {
            generateCombs(openUsed + 1, closedUsed, n, ans, curr + '(');
        }
        if(closedUsed < openUsed) {
            generateCombs(openUsed, closedUsed + 1, n, ans, curr + ')');
        }
    }
}

File: 0022-generate-parentheses/solution.swift
class Solution {
    func generateParenthesis(_ n: Int) -> [String] {
        var ans = [String]()
        var res = ""
        
        solve(0,0)

        // 1. What is the base case?
        // 2. When can I add "("?
        // 3. When can I add ")"?
        func solve(_ open: Int, _ closed: Int) {
            // base case
            if open == n && closed == n {
                ans.append(res)
                return
            }


            // check whether choice is allowed
            if open < n {
                // choose
                res.append("(")
                // recurse
                solve(open+1, closed)
                // undo
                res.removeLast()
            }
            
            // check whether choice is allowed, without open there is no closed
            if closed < open {
                // choose
                res.append(")")
                // recurse
                solve(open, closed+1)
                // undo
                res.removeLast()
            }
        }
        return ans
    }
}
